File: src/utils/formatting.ts
/**
 * Telegram text formatting and MarkdownV2 / HTML escaping utilities.
 *
 * @packageDocumentation
 */

const MARKDOWN_V2_RESERVED_CHARS = /[_*[\]()~`>#+\-=|{}.!\\/]/g;

/**
 * Escapes special characters for Telegram's MarkdownV2 parse mode.
 *
 * @param text - Raw unescaped string.
 * @returns Safely escaped string for MarkdownV2 formatting.
 *
 * @example
 * ```ts
 * const safeText = escapeMarkdownV2("Price: $10.00! [Sale]");
 * // Returns: "Price: $10\\.00\\! \\[Sale\\]"
 * ```
 */
export function escapeMarkdownV2(text: string): string {
  return text.replace(MARKDOWN_V2_RESERVED_CHARS, "\\$&");
}

/**
 * Escapes HTML entities (`&`, `<`, `>`, `"`, `'`) for Telegram's HTML parse mode.
 *
 * @param text - Raw unescaped string.
 * @returns Safely escaped string for HTML formatting.
 *
 * @example
 * ```ts
 * const safeHtml = escapeHtml("<script>alert('xss') & win</script>");
 * // Returns: "&lt;script&gt;alert(&#39;xss&#39;) &amp; win&lt;/script&gt;"
 * ```
 */
export function escapeHtml(text: string): string {
  return text.replace(/[&<>"']/g, (char) => {
    switch (char) {
      case "&":
        return "&amp;";
      case "<":
        return "&lt;";
      case ">":
        return "&gt;";
      case '"':
        return "&quot;";
      case "'":
        return "&#39;";
      default:
        return char;
    }
  });
}

/**
 * Wraps text with bold styling in either HTML or MarkdownV2 format.
 *
 * @param text - Content to format (will be automatically escaped).
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted bold text string.
 */
export function bold(text: string, mode: "HTML" | "MarkdownV2" = "HTML"): string {
  if (mode === "HTML") {
    return `<b>${escapeHtml(text)}</b>`;
  }
  return `*${escapeMarkdownV2(text)}*`;
}

/**
 * Wraps text with italic styling in either HTML or MarkdownV2 format.
 *
 * @param text - Content to format (will be automatically escaped).
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted italic text string.
 */
export function italic(text: string, mode: "HTML" | "MarkdownV2" = "HTML"): string {
  if (mode === "HTML") {
    return `<i>${escapeHtml(text)}</i>`;
  }
  return `_${escapeMarkdownV2(text)}_`;
}

/**
 * Wraps text in an inline code or multi-line preformatted code block.
 *
 * @param text - Code text (will be escaped).
 * @param language - Optional syntax highlighting language (e.g. `"typescript"`).
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted code block string.
 */
export function code(
  text: string,
  language?: string,
  mode: "HTML" | "MarkdownV2" = "HTML",
): string {
  if (language) {
    if (mode === "HTML") {
      return `<pre><code class="language-${escapeHtml(language)}">${escapeHtml(text)}</code></pre>`;
    }
    return `\`\`\`${escapeMarkdownV2(language)}\n${text.replace(/[`\\]/g, "\\$&")}\n\`\`\``;
  }

  if (mode === "HTML") {
    return `<code>${escapeHtml(text)}</code>`;
  }
  return `\`${text.replace(/[`\\]/g, "\\$&")}\``;
}

/**
 * Formats a clickable URL hyperlink.
 *
 * @param text - Label text to display.
 * @param url - HTTPS or tg:// target URL.
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted hyperlink string.
 */
export function link(text: string, url: string, mode: "HTML" | "MarkdownV2" = "HTML"): string {
  if (mode === "HTML") {
    return `<a href="${escapeHtml(url)}">${escapeHtml(text)}</a>`;
  }
  return `[${escapeMarkdownV2(text)}](${escapeMarkdownV2(url)})`;
}

/**
 * Formats an inline mention to a Telegram user by their ID.
 *
 * @param name - Display name of the user.
 * @param userId - Unique Telegram user ID.
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted mention string.
 */
export function userMention(
  name: string,
  userId: number,
  mode: "HTML" | "MarkdownV2" = "HTML",
): string {
  return link(name, `tg://user?id=${userId}`, mode);
}

/**
 * Wraps text with spoiler concealment formatting.
 *
 * @param text - Content to conceal behind spoiler.
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted spoiler string.
 */
export function spoiler(text: string, mode: "HTML" | "MarkdownV2" = "HTML"): string {
  if (mode === "HTML") {
    return `<tg-spoiler>${escapeHtml(text)}</tg-spoiler>`;
  }
  return `||${escapeMarkdownV2(text)}||`;
}

/**
 * Wraps text with strikethrough styling.
 *
 * @param text - Content to format.
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted strikethrough string.
 */
export function strikethrough(text: string, mode: "HTML" | "MarkdownV2" = "HTML"): string {
  if (mode === "HTML") {
    return `<s>${escapeHtml(text)}</s>`;
  }
  return `~${escapeMarkdownV2(text)}~`;
}

/**
 * Wraps text with underline styling.
 *
 * @param text - Content to format.
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted underline string.
 */
export function underline(text: string, mode: "HTML" | "MarkdownV2" = "HTML"): string {
  if (mode === "HTML") {
    return `<u>${escapeHtml(text)}</u>`;
  }
  return `__${escapeMarkdownV2(text)}__`;
}

/**
 * Formats a blockquote quotation.
 *
 * @param text - Quoted content.
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted blockquote string.
 */
export function blockquote(text: string, mode: "HTML" | "MarkdownV2" = "HTML"): string {
  if (mode === "HTML") {
    return `<blockquote>${escapeHtml(text)}</blockquote>`;
  }
  return `>${escapeMarkdownV2(text)}`;
}

/**
 * Formats an expandable collapsible blockquote quotation.
 *
 * @param text - Quoted content.
 * @param mode - Target parse mode (`"HTML"` or `"MarkdownV2"`). Default: `"HTML"`.
 * @returns Formatted expandable blockquote string.
 */
export function expandableBlockquote(text: string, mode: "HTML" | "MarkdownV2" = "HTML"): string {
  if (mode === "HTML") {
    return `<blockquote expandable>${escapeHtml(text)}</blockquote>`;
  }
  return `**>${escapeMarkdownV2(text)}||`;
}

File: packages/go/pkg/utils/formatting.go
// Package utils provides shared helpers for Telegram text formatting
// (MarkdownV2 / HTML escaping and entity builders), input validation
// (bot tokens and Mini App initData), and rate limiting (token bucket).
package utils

import (
	"strconv"
	"strings"
)

// Mode identifies a Telegram parse mode for formatted text.
type Mode string

// Supported parse modes.
const (
	// ModeHTML selects Telegram's HTML parse mode.
	ModeHTML Mode = "HTML"
	// ModeMarkdownV2 selects Telegram's MarkdownV2 parse mode.
	ModeMarkdownV2 Mode = "MarkdownV2"
)

// mdv2ReservedChars lists every character that must be escaped outside
// pre-formatted code spans in Telegram's MarkdownV2 parse mode.
const mdv2ReservedChars = `_*[]()~` + "`" + `>#+-=|{}.!\/`

// EscapeMarkdownV2 escapes all special characters so that text is rendered
// literally under Telegram's MarkdownV2 parse mode.
func EscapeMarkdownV2(text string) string {
	var b strings.Builder
	b.Grow(len(text))
	for i := 0; i < len(text); i++ {
		c := text[i]
		if strings.IndexByte(mdv2ReservedChars, c) >= 0 {
			b.WriteByte('\\')
		}
		b.WriteByte(c)
	}
	return b.String()
}

// EscapeHTML escapes the HTML entities &, <, >, " and ' so that text is
// rendered literally under Telegram's HTML parse mode.
func EscapeHTML(text string) string {
	var b strings.Builder
	b.Grow(len(text))
	for _, r := range text {
		switch r {
		case '&':
			b.WriteString("&amp;")
		case '<':
			b.WriteString("&lt;")
		case '>':
			b.WriteString("&gt;")
		case '"':
			b.WriteString("&quot;")
		case '\'':
			b.WriteString("&#39;")
		default:
			b.WriteRune(r)
		}
	}
	return b.String()
}

// escapeCodeMarkdownV2 escapes only the characters that terminate code
// spans in MarkdownV2 (backtick and backslash); inside code spans the other
// reserved characters must stay literal.
func escapeCodeMarkdownV2(text string) string {
	var b strings.Builder
	b.Grow(len(text))
	for i := 0; i < len(text); i++ {
		c := text[i]
		if c == '`' || c == '\\' {
			b.WriteByte('\\')
		}
		b.WriteByte(c)
	}
	return b.String()
}

// Bold wraps text with bold styling. The content is escaped automatically.
// Any mode other than ModeHTML is treated as MarkdownV2.
func Bold(text string, mode Mode) string {
	if mode == ModeHTML {
		return "<b>" + EscapeHTML(text) + "</b>"
	}
	return "*" + EscapeMarkdownV2(text) + "*"
}

// Italic wraps text with italic styling. The content is escaped automatically.
// Any mode other than ModeHTML is treated as MarkdownV2.
func Italic(text string, mode Mode) string {
	if mode == ModeHTML {
		return "<i>" + EscapeHTML(text) + "</i>"
	}
	return "_" + EscapeMarkdownV2(text) + "_"
}

// Code wraps text as inline code, or as a fenced code block when language is
// non-empty (enabling syntax highlighting for that language). The content is
// escaped automatically. Any mode other than ModeHTML is treated as MarkdownV2.
func Code(text, language string, mode Mode) string {
	if language != "" {
		if mode == ModeHTML {
			return `<pre><code class="language-` + EscapeHTML(language) + `">` + EscapeHTML(text) + "</code></pre>"
		}
		return "```" + EscapeMarkdownV2(language) + "\n" + escapeCodeMarkdownV2(text) + "\n```"
	}
	if mode == ModeHTML {
		return "<code>" + EscapeHTML(text) + "</code>"
	}
	return "`" + escapeCodeMarkdownV2(text) + "`"
}

// Link formats a clickable hyperlink. The label and URL are escaped
// automatically. Any mode other than ModeHTML is treated as MarkdownV2.
func Link(text, url string, mode Mode) string {
	if mode == ModeHTML {
		return `<a href="` + EscapeHTML(url) + `">` + EscapeHTML(text) + "</a>"
	}
	return "[" + EscapeMarkdownV2(text) + "](" + EscapeMarkdownV2(url) + ")"
}

// UserMention formats an inline mention of a Telegram user by their ID using
// a tg://user deep link. Any mode other than ModeHTML is treated as MarkdownV2.
func UserMention(name string, userID int64, mode Mode) string {
	return Link(name, "tg://user?id="+strconv.FormatInt(userID, 10), mode)
}

// Spoiler wraps text with spoiler concealment formatting. The content is
// escaped automatically. Any mode other than ModeHTML is treated as MarkdownV2.
func Spoiler(text string, mode Mode) string {
	if mode == ModeHTML {
		return "<tg-spoiler>" + EscapeHTML(text) + "</tg-spoiler>"
	}
	return "||" + EscapeMarkdownV2(text) + "||"
}

// Strike wraps text with strikethrough styling. The content is escaped
// automatically. Any mode other than ModeHTML is treated as MarkdownV2.
func Strike(text string, mode Mode) string {
	if mode == ModeHTML {
		return "<s>" + EscapeHTML(text) + "</s>"
	}
	return "~" + EscapeMarkdownV2(text) + "~"
}

// Underline wraps text with underline styling. The content is escaped
// automatically. Any mode other than ModeHTML is treated as MarkdownV2.
func Underline(text string, mode Mode) string {
	if mode == ModeHTML {
		return "<u>" + EscapeHTML(text) + "</u>"
	}
	return "__" + EscapeMarkdownV2(text) + "__"
}

// Blockquote formats text as a quotation. The content is escaped
// automatically. Any mode other than ModeHTML is treated as MarkdownV2.
func Blockquote(text string, mode Mode) string {
	if mode == ModeHTML {
		return "<blockquote>" + EscapeHTML(text) + "</blockquote>"
	}
	return ">" + EscapeMarkdownV2(text)
}

// ExpandableBlockquote formats text as a collapsed, expandable quotation.
// The content is escaped automatically. Any mode other than ModeHTML is
// treated as MarkdownV2.
func ExpandableBlockquote(text string, mode Mode) string {
	if mode == ModeHTML {
		return "<blockquote expandable>" + EscapeHTML(text) + "</blockquote>"
	}
	return "**>" + EscapeMarkdownV2(text) + "||"
}
